From 102ebb4561f48c48f0bfc3ab5d6922291285f03f Mon Sep 17 00:00:00 2001
From: Ambroise Vincent <ambroise.vincent@arm.com>
Date: Mon, 10 Jul 2023 10:38:05 +0100
Subject: [PATCH] shell: Provide a configuration to pin the shell thread

Introduce a CONFIG_SHELL_PIN_THREAD configuration options to pin the
shell thread to CPU 0.

Signed-off-by: Ambroise Vincent <ambroise.vincent@arm.com>
Upstream-Status: Inappropriate [Kronos-specific]
---
 subsys/shell/Kconfig |  7 +++++++
 subsys/shell/shell.c | 12 +++++++++++-
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/subsys/shell/Kconfig b/subsys/shell/Kconfig
index 43035e520e..3daa7061f1 100644
--- a/subsys/shell/Kconfig
+++ b/subsys/shell/Kconfig
@@ -268,6 +268,13 @@ config SHELL_CMDS_RETURN_VALUE
 	  This option enables the retval command. It is used to retrieve
 	  the return value from the most recently executed command.
 
+config SHELL_PIN_THREAD
+	bool "Pin thread"
+	default n
+	depends on SCHED_CPU_MASK
+	help
+	  When enabled, the shell thread will be pinned to CPU 0.
+
 source "subsys/shell/modules/Kconfig"
 
 endif # SHELL
diff --git a/subsys/shell/shell.c b/subsys/shell/shell.c
index 34ed7daa75..8f33442707 100644
--- a/subsys/shell/shell.c
+++ b/subsys/shell/shell.c
@@ -1387,7 +1387,17 @@ int shell_init(const struct shell *sh, const void *transport_config,
 				  sh->stack, CONFIG_SHELL_STACK_SIZE,
 				  shell_thread, (void *)sh, (void *)log_backend,
 				  UINT_TO_POINTER(init_log_level),
-				  SHELL_THREAD_PRIORITY, 0, K_NO_WAIT);
+				  SHELL_THREAD_PRIORITY, 0, K_FOREVER);
+
+#if defined(CONFIG_SHELL_PIN_THREAD)
+	err = k_thread_cpu_pin(tid, 0);
+	if (err) {
+		z_shell_fprintf(sh, SHELL_WARNING,
+				"Failed to pin shell thread: %d\n", err);
+	}
+#endif /* CONFIG_SHELL_PIN_THREAD */
+
+	k_thread_start(tid);
 
 	sh->ctx->tid = tid;
 	k_thread_name_set(tid, sh->thread_name);
-- 
2.25.1

